### ISSUE TITLE
Generate Minimal View for Domino Visual Editor Agents

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1073]
In order for the agents generated by #1072 to work, we need a view for each form that shows only the documents for that form.  

For forms created in Domino Visual Editor, we'll need to generate this view in Moonshine.  For an NSFConverter database, we'll need to generate it either 
- In NSFConverterPortal.  This is a little more intuitive, but requires updates (and template maintenance) in multiple projects.
- In Moonshine (when opening the database)

We'll also need to update or recreate the view if the form is renamed.

I created a template for the view, which contains:
- A selection formula to select only a specific form
- UNID column - so that the view has visible data
- Create Action - convenience tool for testing the database.  This should be generally useful for Domino Visual Editor.

To use this, replace `%form%` in the name and content with the corresponding Form.  Note that the file name will contain spaces - I'm following a Prominic convention for this view.  Do **not** remove spaces and special characters like we do for #1071 - this expects to use the original form name.  The form name should the XML-escaped for the content, though.

UPDATE:  I created and tested the view template:  https://static.moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_978_CRUD_agents_integration/All%20By%20UNID_5cCRUD_5c%25form%25.view
UPDATE2:  I updated the above template for the new subfolder.

**Comment by JoelProminic** [comment]
@JustinProminic requested that we put these minimal views in a subfolder under "All by UNID" so that we can keep them separate from the original views.  This should just be an update to the generated view name.

**Comment by JoelProminic** [comment]
Sorry, I forgot that I still needed to revisit the directory change from my last comment.  Let's update the view names to `All By UNID\CRUD\%form%`.  So, this will let us keep the views more organized if we copy them to a database that already has `All By UNID` views.  The file names will look like this:

    All By UNID_5cCRUD_5c%form%.view

Since my earlier comment was unclear, @feather812002 added a subdirectory under `odp/Views`.  For future reference, I see that this triggers errors in the NSFODP build:

    : [INFO] Caused by: java.lang.RuntimeException: Exception while importing element Views/All by UNID
    ....
    : [INFO] Caused by: java.lang.RuntimeException: java.io.IOException: Is a directory

**Comment by JoelProminic** [comment]
Some other notes from today's discussion:

I see that if we put an underscore in the view name, it is cleared in the actual generated view:
![image](https://user-images.githubusercontent.com/24532731/187969810-b76e7b3d-46f9-4ec4-98f5-3e67fa0bdcec.png)

This makes sense, since the `\` is encoded as `_5c`.  I tried encoding the underscore as `_5f`, but it didn't work.  This needs more investigation.

We will also need to support renaming these .view files when the forms are renamed.  The form name will also need to be updated in the .view file contents, but this should be a simple search and replace.

**Comment by feather812002** [comment]
@JoelProminic  I already update the View template to latest version .
Now, on the Moonshine-IDE and NSFConvertProtal both all should work on the last view template:

![image](https://user-images.githubusercontent.com/8543391/188636781-9818c432-bf6f-46f0-8d31-823147d889d3.png)

I also to do some test about this , I confirm the export project from NSFConvertProtal can work fine with Moonshine-IDE
as you see on above picture , the latest view show the CURD as sub  folder .

**Comment by feather812002** [comment]
I also merge the latest update from master to branch , so if all is ready , I think this week we can try merge the branch into master. let me know if you found other some bug or issue ,

**Comment by feather812002** [comment]
@JoelProminic 
I also update the rename for simple view, if user rename a form name , Moonshine-IDE will auto search the source simple view and replace the from name as new name .
I will submit this, in my first test it work fine .
![image](https://user-images.githubusercontent.com/8543391/188651853-2686e6dc-5e34-477a-84b2-16cd118a1da5.png)

**Comment by JoelProminic** [comment]
@JustinProminic noticed today, that we have an entry in the File > New menu for this view template.

![image](https://user-images.githubusercontent.com/24532731/192650462-ed4ea889-eecf-4fd0-8ee5-103163578b9f.png)

I tried creating a new file from this, and I found it appended the name to the beginning of the template and did not give me an option to replace "%form%"

![image](https://user-images.githubusercontent.com/24532731/192650373-36e6802f-9a0d-4de8-9448-6dd5b91bc2f4.png)

I don't think we need or want this for now.    If we need this to be in the file template list, I'd like to hide it from the menu.

**Comment by JoelProminic** [comment]
I also see that the views were not deleted when the corresponding form was deleted.   In theory the old views could be useful for "legacy" operations, but I don't think that makes sense for our use cases:
![image](https://user-images.githubusercontent.com/24532731/192650775-53c04979-2f2b-47ba-a1ce-d8afe8f14380.png)

**Comment by JoelProminic** [comment]
I also had trouble with the view names in a project with underscores in the name "TestDominoVisualEditor_0923".   I found that the agent needed to exclude the underscore in order to open the view.

```
		return agentDatabase.getView("All By UNID/CRUD/TestDominoVisualEditor0923");
		// return agentDatabase.getView("All By UNID/CRUD/TestDominoVisualEditor_0923"); // expected
```

In my earlier tests for this issue, the underscores were preserved, but this result seems to contradict this finding.  I was seeing other strange behavior with this test, so it needs more investigation.

**Comment by feather812002** [comment]
@JoelProminic  I already fix the "All By UNID_5cCRUD_5c%form%" appear in the "File-> New" menu.

![image](https://user-images.githubusercontent.com/8543391/193022615-91b49c55-b6aa-4408-a2f2-5ff4bf63663a.png)


But for the old simple view for already remove form , let me clear why I keep it .

As we know , the view formula can contain more than one form , if we don't want user edit the simple view 
we force the simple view only work for one from ,and it is fixed and can't be change , we need remove the view after the form be remove, this is fine .

So, I want to confirm this , above description logic is meet your requirements?

**Comment by feather812002** [comment]
@JoelProminic 
About the underscore , I can't repeat your bug , I always get the view have the correct name.
![image](https://user-images.githubusercontent.com/8543391/193022988-44c9d1eb-0097-40ca-9ca0-0de9aa215333.png)

Can you details how do you trigger it ?

Thanks.

**Comment by JoelProminic** [comment]
I confirmed that the CRUD views are getting deleted on Form deletion with the latest update.  I also confirmed this logic does not trigger an error if the view does not exist.  I also checked that they are still renamed when the Form is renamed as well.

The template has been removed from the File > New and Right-Click > New interfaces.

The underscore issue triggers when I try to open the view from a Java agent.  I still need to investigate this, but it may require changes to the naming convention for these views.  Alternatively, we may end up changing the Java generation or template instead.
